Fixing low contrast between light terminal backgrounds with program font colors.

saguileran opened the pull request:
Hi Luan.

The additional code added implements a method to verify the terminal background color based on the operating system. Then, based on the resulting mean of the terminal background colors a colors set is defined between a dark and light themes, where the colors are opposed.

Let me know if the code follows your code style and if the colors chosen are suitable.

Commit:
Fixing low contrast between light terminal backgrounds with program font colors. The aditional code
added implements a method to verify the terminal background color based on the oeprating system.
Then, based on the resulting mean of the terminal background colors a colors set is defined between a
dark and light themes, where the colors are opposed.

LipArcanjo commented on arkanjo/base/utils.cpp:
```diff
@@ -124,3 +130,130 @@ vector<string> Utils::split_string(string s, char delimiter){
+  return mean_color;
+}
+
+bool Utils::is_bg_color_dark(){
```
Can you create three utils .h for the OS? Like windows_utils.h, linux_utils.h and apple_utils.h. So we can encapsulate the operation system logic. 
On the header file of the utils we import the specific .h. 

As most of the logic repeats between linux and apple, you can also create a apple_linux_utils.h to the shared code. It should have better ways to handle that, but let us not over engineering on a couple of functions.

Just define in all of them the namespace UtilsOSDependable or something like that and implement the is_bg_color_dark.

We will have other changes that depends on the operationg system and I want to concentrate this kind of logic in a one place to know where to look.

saguileran replied:
Ok, I will do that. However, I did not understand all your comment.

You are saying to create three new files: windows_utils.h, linux_utils.h, and apple_utils.h, but I am a little confused about the extension, can be them .hpp or they have to be .h? 

In my current solution, I am creating three files: windows_utils.hpp, linux_utils.hpp, and apple_utils.hpp. Then, for testing porpuse, I am calling the corerct one with:

```c++
#ifdef _WIN32
    #include <windows.h>
		#include "windows_utils.hpp"
#elif __linux__
    #include <sys/utsname.h>
		#include "linux_utils.hpp"
#elif __APPLE__
    #include <TargetConditionals.h>
		#include "apple_utils.hpp" 
#endif
```

Is it correct? I know you said you are going to do it but I need this part to test the cange in all systems.

In all three headers I am creating the name space `UtilsOSDependable` with a `is_bg_color_dark()` function. Nevertheless, I do not know how to call the function from `utils.cpp`, could it be something like: `UtilsOSDependable::is_bg_color_dark()`? I did it in this way but I am getting the following error:

```bash
na função "Utils::format_colored_message(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, Utils::COLOR)":
../arkanjo/base/utils.cpp:91:(.text+0x918): undefined reference to `UtilsOSDependable::is_bg_color_dark()'
collect2: error: ld returned 1 exit status
```

Do you know how can I solve it? I think it is about the namespace importing but I am not sure.

Related to other OS, I have not tested yet. Howerver, I will do it after the code is in the suitable structure you mentioned.

LipArcanjo replied:
I mean to create hpp in fact. Sorry for the confusion.

Just a sanity check, have you added the .cpp equivalent files and have added the targets to the Makefile? I do not know why it is not working. Maybe the C++ linker does not let this to work.

Can you create a commit in your fork with the new changes so I can test myself what is happening?

Btw if none of this works. it will be fine to keep it as is. Just create three functions for each of the behaviours if possible.

Thanks

LipArcanjo replied:
Why `#include <windows.h>` cannot be inside of `windows_utils.hpp`  ?

Also, about the error. We may try to add the `#ifdef _WIN32` to the whole `windows_utils.hpp`, so the compiler preprocessor will remove everything, so it will not have multiple definitions of the function. If this is in fact the problem.

Same logic for the other OS

LipArcanjo commented on arkanjo/base/utils.hpp:
```diff
@@ -8,6 +8,26 @@
+    #include <sys/utsname.h>
+#elif __APPLE__
+    #include <TargetConditionals.h>
+#endif
```
Related to my comment on `is_bg_color_dark`. Let us just create a utils for each OS and let the specific utils import the required libraries.

LipArcanjo commented on arkanjo/base/utils.cpp:
```diff
@@ -77,7 +77,13 @@ bool Utils::is_regular_file(string path){
 
 string Utils::format_colored_message(string message, COLOR color){
-	return COLOR_TOKENS_UTILS[color] + message + COLOR_TOKENS_UTILS[RESET];
+	if (is_bg_color_dark()){
```
Not required for you to do, but would be nice to have a option to not print colored output at all. 

You can use the config.h to keep a boolean and call that here. in orchestration.h you can set the boolean given a passed parameter.

If you do not want to do that, it is fine :)

LipArcanjo commented on arkanjo/base/utils.hpp:
```diff
@@ -8,6 +8,26 @@
+#include <sstream>
+#include <iomanip>
+
+#ifdef _WIN32
```
Have you tested the logic on windows? If not, add comments that has not been properly tested yet.

LipArcanjo commented on arkanjo/base/utils.hpp:
```diff
@@ -8,6 +8,26 @@
+#ifdef _WIN32
+    #include <windows.h>
+#elif __linux__
+    #include <sys/utsname.h>
```
Have you tested the logic on MacOS? If not, add comments that has not been properly tested yet.

LipArcanjo commented on arkanjo/base/utils.hpp:
```diff
@@ -59,7 +93,16 @@ namespace Utils{
+  bool is_bg_dark();
+  string capture_terminal_response();
+  bool is_bg_color_dark();
+	int get_terminal_bg_color_mean();
```
fix indentation typo.

LipArcanjo reviewed (changes requested):
Hi Sebastian. Thanks for the pull request!

I have added some comments.

I will test the changes tomorrow to test if the light and dark mode is working as intended. But you can start working on the comments.

Commit:
Update of the code creating three headers for each OS. The code is tested on linux and windows, apple testing is not done yet.
The headers share some functions and the corresponding header, depending on the OS, is called on the utils.h header.
To calculate is the terminal background color is light or dark, the luminance is computed using the rgb color.

saguileran committed:
Create main.yml

saguileran committed:
Add files via upload

saguileran committed:
Update Doxyfile

Commit:
Ignoring docs and vscode configurations

Commit:
Fixing gh-pages doxygen deployment

Commit:
Fixing worjflow file

Commit:
V2 fixing

Commit:
Update Doxyfile and github action workflow file

Commit:
Update readme to display on table of toc and doxyfile
